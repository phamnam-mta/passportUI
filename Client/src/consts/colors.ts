import { SharedColors } from "@fluentui/theme";

export const colors = [
    SharedColors.red10,
    SharedColors.orange10,
    SharedColors.yellowGreen10,
    SharedColors.cyan30,
    SharedColors.cyanBlue10,
    SharedColors.blueMagenta10,
    SharedColors.magentaPink10,
    SharedColors.red20,
    SharedColors.orange20,
    SharedColors.orangeYellow10,
    SharedColors.greenCyan10,
    SharedColors.cyanBlue20,
    SharedColors.blueMagenta20,
    SharedColors.magentaPink20,
    SharedColors.redOrange10,
    SharedColors.yellow10,
    SharedColors.green20,
    SharedColors.cyan20,
    SharedColors.blue10,
    SharedColors.magenta20,
    SharedColors.pinkRed10,
    SharedColors.orange30,
    SharedColors.orangeYellow20,
    SharedColors.green10,
    SharedColors.cyan10,
    SharedColors.blueMagenta30,
    SharedColors.magenta10,
];
