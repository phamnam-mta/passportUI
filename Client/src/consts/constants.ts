export const constants = {
    defaultSplitPaneSizes: {
        analyzeSplitPaneSize: [70, 30],
        labelSplitPaneSize: [80, 20],
        labelTableSplitPaneSize: [65, 35],
    },
    dynamicTableImgSrc: "/images/customModels/dynamic-table.png",
    fixedTableImgSrc: "/images/customModels/fixed-table.png",
    fieldsSchema: "https://schema.cognitiveservices.azure.com/formrecognizer/2021-03-01/fields.json",
    labelsSchema: "https://schema.cognitiveservices.azure.com/formrecognizer/2021-03-01/labels.json",
    fieldsFile: "fields.json",
    labelFileExtension: ".labels.json",
    ocrFileExtension: ".ocr.json",
};

export enum LoadingOverlayWeights {
    ExtraLight = 0,
    Light = 10,
    Default = 20,
    SemiHeavy = 30,
    Heavy = 40,
    ExtraHeavy = 50,
}

export enum KeyEventType {
    KeyDown = "keydown",
    KeyUp = "keyup",
}

export enum KeyEventCode {
    Shift = "Shift",
    Escape = "Escape",
}
