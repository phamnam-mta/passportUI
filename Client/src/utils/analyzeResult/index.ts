import { BoundingRegion, Polygon, StudioFieldValue, StudioSpan } from "models/analyzeResult";

export const getPagePolygons = (boundingRegions: BoundingRegion[], page: number): Polygon[] => {
    return boundingRegions
        .filter((boundingRegion) => boundingRegion.pageNumber === page)
        .map((boundingRegion) => boundingRegion.polygon);
};

export const isAnyBoundingRegionInPage = (boundingRegions: BoundingRegion[], page: number): boolean => {
    return boundingRegions.some((boundingRegion) => boundingRegion.pageNumber === page);
};

export const getFieldTypeValue = (field: StudioFieldValue) => {
    const { type } = field;
    const typeString = type.charAt(0).toUpperCase() + type.slice(1);

    // TODO: Update when going public.
    if (type === "currency" && field.valueCurrency && typeof field.valueCurrency !== "string") {
        const { currencySymbol, currency, amount } = field.valueCurrency;
        if (currencySymbol && amount) {
            return `${currencySymbol}${amount}`;
        } else if (currency && amount) {
            return `${currency}${amount}`;
        } else if (amount) {
            return `${amount}`;
        } else {
            return undefined;
        }
    } else if (type === "address" && field.valueAddress) {
        return field.content;
    }
    return field[`value${typeString}`];
};

export interface SpanPosition extends StudioSpan {
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
    shouldDisplayAsIcon: boolean;
}
