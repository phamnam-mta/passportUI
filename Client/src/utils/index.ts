export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
