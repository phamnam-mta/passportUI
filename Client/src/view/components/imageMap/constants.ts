export const LAYER_NAME = "name";
export const IMAGE_LAYER_NAME = "imageLayer";
export const TEXT_VECTOR_LAYER_NAME = "textVectorLayer";
export const POD_VECTOR_LAYER_NAME = "podVectorLayer";
export const TABLE_BORDER_VECTOR_LAYER_NAME = "tableBorderVectorLayer";
export const TABLE_ICON_VECTOR_LAYER_NAME = "tableIconVectorLayer";
export const TABLE_ICON_BORDER_VECTOR_LAYER_NAME = "tableIconBorderVectorLayer";
export const CHECKBOX_VECTOR_LAYER_NAME = "checkboxBorderVectorLayer";
export const LABEL_VECTOR_LAYER_NAME = "labelledVectorLayer";
export const DRAWN_REGION_LABEL_VECTOR_LAYER_NAME = "drawnRegionLabelledVectorLayer";
export const DRAWN_REGION_VECTOR_LAYER_NAME = "drawnRegionVectorLayer";

export const SELECTED_PROPERTY = "selected";
export const SELECTION_MARK_STATE = "selectionMarkState";
export const HIGHLIGHTED_PROPERTY = "highlighted";
export const FIELD_PROPERTY = "field";
export const COLOR_PROPERTY = "color";
export const BACKGROUND_COLOR_PROPERTY = "backgroundColor";
export const DASHED_PROPERTY = "dashed";
export const DISABLEHIGHLIGHT_PROPERTY = "disableHighlight";
