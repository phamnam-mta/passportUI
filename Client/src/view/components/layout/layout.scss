.main {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    height: 100%;

    .page-container {
        display: flex;
        flex-grow: 1;
    }
}
