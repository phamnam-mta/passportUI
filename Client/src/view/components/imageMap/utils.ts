import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
