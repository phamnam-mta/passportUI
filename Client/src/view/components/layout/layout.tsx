import * as React from "react";
import { connect, ConnectedProps } from "react-redux";
import Footer from "../footer/footer";
import { ApplicationState } from "store";
import LoadingOverlay from "../loadingOverlay/loadingOverlay";
import { ILoadingOverlay } from "store/portal/portal";

import "./layout.scss";

interface ILayoutProps {
    children: React.ReactNode;
}
type Props = ConnectedProps<typeof connector> & ILayoutProps;

export class Layout extends React.PureComponent<Props> {
    private getLoadingOverlay = (): ILoadingOverlay => {
        const { loadingOverlays } = this.props;
        const sortedLoadingOverlay = [...loadingOverlays].sort(({ weight: a }, { weight: b }) => b - a);
        return sortedLoadingOverlay[0];
    };

    public render() {
        const { loadingOverlays } = this.props;
        const isLoading = loadingOverlays.length > 0;

        return (
            <React.Fragment>
                <div className="main" aria-busy={isLoading ? true : false} aria-hidden={isLoading ? true : false}>
                    <div role="status" className="sr-only">
                        {isLoading ? (
                            <span key="loading">Page is loading</span>
                        ) : (
                            <span key="loaded">Page is loaded</span>
                        )}
                    </div>
                    <div className="page-container">{this.props.children}</div>
                    <Footer />
                </div>
                {isLoading && <LoadingOverlay message={this.getLoadingOverlay().message} />}
            </React.Fragment>
        );
    }
}

const mapState = (state: ApplicationState) => ({
    location: state.router.location,
    loadingOverlays: state.portal.loadingOverlays,
});

const connector = connect(mapState);

export default connector(Layout);
