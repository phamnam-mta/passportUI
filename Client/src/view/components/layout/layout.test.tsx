import { LoadingOverlayWeights } from "consts/constants";
import { shallow } from "enzyme";
import * as React from "react";
import { mockLoadingOverlays } from "utils/test";
import { Layout } from "./layout";

describe("<Layout />", () => {
    let baseProps;

    beforeEach(() => {
        baseProps = {
            location: {
                pathname: "/studio/read",
            },
            loadingOverlays: [],
        };
    });

    describe("Rendering", () => {
        it("should render correctly when it's on homePage", () => {
            const props = {
                ...baseProps,
                location: {
                    pathname: "/studio",
                },
            };
            const wrapper = shallow(<Layout {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly when there is a loading layer on Layout", () => {
            const props = {
                ...baseProps,
                loadingOverlays: mockLoadingOverlays,
            };
            const wrapper = shallow(<Layout {...props} />);
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe("Event Handling", () => {
        it("should render & sort loading overlay correctly when there are multiple loading layers on Layout", () => {
            const props = {
                ...baseProps,
                loadingOverlays: mockLoadingOverlays,
            };
            const wrapper = shallow(<Layout {...props} />);
            const instance = wrapper.instance() as any;
            expect(instance.getLoadingOverlay().weight).toBe(LoadingOverlayWeights.Heavy);
        });
    });
});
