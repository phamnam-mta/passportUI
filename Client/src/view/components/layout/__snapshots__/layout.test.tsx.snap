// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> Rendering should render correctly when it's on homePage 1`] = `
<Fragment>
  <div
    aria-busy={false}
    aria-hidden={false}
    className="main"
  >
    <div
      className="sr-only"
      role="status"
    >
      <span
        key="loaded"
      >
        Page is loaded
      </span>
    </div>
    <div
      className="page-container"
    />
    <Footer />
  </div>
</Fragment>
`;

exports[`<Layout /> Rendering should render correctly when there is a loading layer on Layout 1`] = `
<Fragment>
  <div
    aria-busy={true}
    aria-hidden={true}
    className="main"
  >
    <div
      className="sr-only"
      role="status"
    >
      <span
        key="loading"
      >
        Page is loading
      </span>
    </div>
    <div
      className="page-container"
    />
    <Footer />
  </div>
  <LoadingOverlay
    message="Loading..."
  />
</Fragment>
`;
